tools/runtest

```diff
@@ -414,158 +248,393 @@ set_qexe_testlist(){
     fi
 
     if [ "$test_length" = 'full' ]; then
-      test_list="$testdir/testlist_full_cuda.txt"
+      test_list="$testdir/testlist_full_gpu.txt"
     else
-      test_list="$testdir/testlist_short_cuda.txt"
+      test_list="$testdir/testlist_short_gpu.txt"
     fi
   fi
+}
+
+
+# function to set the GPUs assigned to GNU parallel jobs
+#
+# arguments:
+#   $1: available GPU IDs (0-based indexing) (i.e., {CUDA,HIP}_VISIBLE_DEVICES)
+# outputs:
+#   total_gpus: total number of GPUs
+#   gpus_per_job: number of GPUs per job
+#   job_gpus: array of GPU IDs assigned to each job (0-based indexing)
+set_job_gpus()
+{
+  total_gpus=`echo "$1" | awk '{print NF}'`
+  gpus_per_job=$((total_gpus/PARALLEL_TEST_COUNT))
+
+  if [ "$total_gpus" -lt "$PARALLEL_TEST_COUNT" ]; then
+    echo "Error: invalid # GPUs detected in HIP_VISIBLE_DEVICES/CUDA_VISIBLE_DEVICES (# GPUs >= PARALLEL_TEST_COUNT)."
+    exit 1;
+  fi
 
+  job_gpus=()
+  for i in `seq 0 $(($PARALLEL_TEST_COUNT-1))`;
+  do
+    job_gpus[$i]="`echo $1 | cut -f $((i*gpus_per_job+1))-$(((i+1)*gpus_per_job)) -d ' '`"
+  done
 }
 
 
-# function to get the total number of tests
-get_total_tests(){
+# function to set the various test lists and counts
+#
+# arguments:
+#   $1: run energy tests ['yes'|'no']
+#   $2: run gradient tests ['yes'|'no']
+#   $3: run geometry optimization tests ['yes'|'no']
+#   $4: run API tests ['yes'|'no']
+#
+# outputs:
+#   tests: array of tests
+#   test_names: array of test names
+#   tot_{ene,grad,opt,api}_tests: counts for test lists by type
+#   total_tests: sum of counts for all test types
+get_total_tests()
+{
+  tests=()
+  test_names=()
+  total_tests=0 
 
-  if [ "$test_ene" = 'yes' ]; then
-    tot_ene_tests=`awk '{print $1}' "$test_list"| grep "ene" | wc -l`
-    total_tests=$((total_tests+tot_ene_tests))
+  if [ "$1" = 'yes' ]; then
+    tot_ene_tests="${#tests[@]}"
+    tests+=(`grep "^ene_" "$test_list" | awk '{print $1}'`)
+    readarray -t -O "${#test_names[@]}" test_names <<< `grep "^ene_" "$test_list" | awk 'BEGIN { FPAT="([^#]+)"; OFS="\n"; } {print $2}'`
```

Comment:
`readarray` is not portable. Behavior changed with bash versions. 

It works as expected on Almalinux 9.4 with bash 
```
[agoetz@flamey install]$ cd $QUICK_INSTALL/test
[agoetz@flamey test]$ echo $BASH_VERSION
5.1.8(1)-release
[agoetz@flamey test]$ readarray -t foo <<<  `grep "^ene_" "testlist_short.txt" | awk 'BEGIN { FPAT="([^#]+)"; OFS="\n"; } {print $2}'`
agoetz@flamey test]$ echo ${#foo[@]}
11
[agoetz@flamey test]$ echo ${foo[0]}
PSB5: RHF energy test: s and p basis functions
```

But this happens on a CentOS 7.9 box with bash 4.2.46.
```
[agoetz@chinotto test]$ cd $QUICK_INSTALL/test
[agoetz@chinotto test]$ echo $BASH_VERSION
4.2.46(2)-release
[agoetz@chinotto test]$ readarray -t foo <<<  `grep "^ene_" "testlist_short.txt" | awk 'BEGIN { FPAT="([^#]+)"; OFS="\n"; } {print $2}'`
[agoetz@chinotto test]$ echo ${#foo[@]}
1
[agoetz@chinotto test]$ echo ${foo[0]}
PSB5: RHF energy test: s and p basis functions Acetone: RHF energy test: s and p basis functions PSB3: DFT energy test: s and p basis functions, BLYP functional PSB3: DFT energy test: s, p and d basis functions, BLYP functional PSB3: DFT energy test: s and p basis functions, native B3LYP functional PSB3: DFT energy test: s, p and d basis functions, native B3LYP functional PSB3: DFT energy test: s and p basis functions, libxc LDA functional PSB3: DFT energy test: s and p basis functions, libxc GGA functional PSB3: DFT energy test: s and p basis functions, libxc hybrid GGA functional Water-2 MP2 energy test: s and p basis functions Water-2 MP2 energy test: s, p and d basis functions
```

Following should be portable. Can you please check?
```
test_names+=(`grep "^ene_" "$test_list" | awk 'BEGIN { FPAT="([^#]+)"; OFS="\n"; } {print $2}'`)
```

Reply:
This enables running tests in serial (without GNU parallel). 

However, when requesting to run tests simultaneously leads to hanging:
```
[agoetz@chinotto install-pr381]$ export PARALLEL_TEST_COUNT=2
[agoetz@chinotto install-pr381]$ ./runtest --serial

Testing QUICK-24.03 serial version started on 10/19/2024 at 08:19:38 PM. 
```
No further output.

Two output files are created in `$QUICK_HOME/test/runs/serial/` corresponding to the first two tests. These tests ran and quit with an error because the input file was not found
```
| Error: Failed to open input file.

| Error Termination. Task Failed on: Sat Oct 19 20:19:38 2024
```
However, no further tests run, which is odd.

Reply:
GNU parallel version on chinotto (CentOS 7.9 box)
```
[agoetz@chinotto install-pr381]$ parallel --version
GNU parallel 20160222
```
Same behavior is expected on RHEL 7, of which there probably still are some around.

Reply:
Actually, the issue is not with readline, but command substitution requires quoting the returned string, otherwise newline characters are stripped, e.g.
```
# This works
readarray -t foo <<< "`grep ...`"
```